Code/GraphMol/SynthonSpaceSearch/SynthonSpaceFingerprintSearcher.cpp

```diff
@@ -0,0 +1,151 @@
+//
+// Copyright (C) David Cosgrove 2024.
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <DataStructs/BitOps.h>
+#include <GraphMol/MolOps.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/SynthonSpaceSearch/SynthonSpaceSearch_details.h>
+#include <GraphMol/SynthonSpaceSearch/SynthonSpaceFingerprintSearcher.h>
+
+namespace RDKit::SynthonSpaceSearch {
+
+SynthonSpaceFingerprintSearcher::SynthonSpaceFingerprintSearcher(
+    const ROMol &query, const FingerprintGenerator<std::uint64_t> &fpGen,
+    const SynthonSpaceSearchParams &params, SynthonSpace &space)
+    : SynthonSpaceSearcher(query, params, space), d_fpGen(fpGen) {
+  if (!getSpace().hasFingerprints() ||
+      getSpace().getSynthonFingerprintType() != fpGen.infoString()) {
+    getSpace().buildSynthonFingerprints(fpGen);
+  }
+  d_queryFP = std::make_unique<ExplicitBitVect>(*d_fpGen.getFingerprint(query));
+}
+
+namespace {
+// Take the fragged mol fps and flag all those synthons that have a fragment as
+// a similarity match.
+std::vector<boost::dynamic_bitset<>> getHitSynthons(
+    const std::vector<std::unique_ptr<ExplicitBitVect>> &fragFPs,
+    const double similarityCutoff, const std::unique_ptr<SynthonSet> &reaction,
+    const std::vector<unsigned int> &synthonOrder) {
+  std::vector<boost::dynamic_bitset<>> synthonsToUse;
+  for (const auto &synthonSet : reaction->getSynthons()) {
+    synthonsToUse.emplace_back(synthonSet.size());
+  }
+  for (size_t i = 0; i < synthonOrder.size(); i++) {
+    const auto &synthonFPs = reaction->getSynthonFPs()[synthonOrder[i]];
+    bool fragMatched = false;
+    for (size_t j = 0; j < synthonFPs.size(); j++) {
+      if (const auto sim = TanimotoSimilarity(*fragFPs[i], *synthonFPs[j]);
```

Comment:
another note for future us: we should make the similarity function that's used configurable

Reply:
Yeah.  It felt complicated so I ducked it.